lib_vehicle_model/src/lib_vehicle_model/ModelLoader.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+/*
+ * Copyright (C) 2018-2019 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <stdexcept>
+#include <vector>
+#include <string>
+#include <dlfcn.h>
+#include <memory>
+#include <stdlib.h>
+#include <sstream>
+#include "lib_vehicle_model/VehicleMotionModel.h"
+
+
+namespace lib_vehicle_model {
+  /**
+   * @class ModelLoader
+   * @brief Class which can load a vehicle model
+   */
+  class ModelLoader
+  {
+    public:
+
+      // Typedef for function pointers to use with loaded libraries create and destroy functions
+      typedef VehicleMotionModel* (*create_fnc_ptr)();
+      typedef void (*destroy_fnc_ptr)(VehicleMotionModel*);
```

Comment:
Can you explain how those two lines of typedef work?

Reply:
I know they are very confusing...

In C++ you can have a pointer to a function
```
// Function
int foo() { return 5;}

// fcnPtr is a pointer to a function that takes no arguments and returns an integer
int (*fcnPtr)();

int (*fcnPtr)() = foo; // fcnPtr points to function foo
fcnPtr(5); // call function foo(5) through fcnPtr.
```
The actual variable name in that example is fcnPtr.

What's confusing about typedef for a function pointer is that the variable name sort of becomes  the type name (unlike how regular type def works...)

Normal typedef
```
typedef int myinteger; // myinteger is the new type so you could say myinteger x = 5;
```
Typedef of function
```
typedef void (*myfunc)(); // myfunc now has a type meaning a pointer to a function which returns void and takes no arguments
myfunc f;      // compile equally as  void (*f)();
```
I got these examples from
https://www.learncpp.com/cpp-tutorial/78-function-pointers/
https://stackoverflow.com/questions/4295432/typedef-function-pointer

Reply:
It looks like in C++11 I could also say. I'm not sure what the limitations of that are, but I
```
using destroy_fnc_ptr = void(*)(VehicleMotionModel*); // type alias
```
I personally find that alot cleaner so I'll look into if that can be used here instead.

Reply:
Great explanations. I will dig into the link you send to me. It is black magic indeed.

Reply:
I think those two lines are pretty clear to me now.